File: 135_Candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        if (n == 0) {
            return 0;
        }

        // Initialize an array to store the candies for each child.
        // Everyone starts with at least one candy.
        int[] candies = new int[n];
        for (int i = 0; i < n; i++) {
            candies[i] = 1;
        }

        // First pass: Left to right
        // If a child has a higher rating than their left neighbor,
        // they get one more candy than their left neighbor.
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1;
            }
        }

        // Second pass: Right to left
        // If a child has a higher rating than their right neighbor,
        // they must get more candies than their right neighbor.
        // We take the maximum of the current candy count and the one based on the right neighbor.
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
            }
        }

        // Sum up all the candies.
        int totalCandies = 0;
        for (int i = 0; i < n; i++) {
            totalCandies += candies[i];
        }

        return totalCandies;
    }
}

File: 135_Candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        // Get the number of children in the line.
        int plen = ratings.size();
        
        // If there are no children, no candies are needed.
        if (plen == 0) return 0;
        
        // Initialize a vector 'treats' to store the number of candies for each child.
        // Every child initially receives at least one candy.
        vector<int> treats(plen, 1);
        
        // --- Left to right pass ---
        // This pass ensures that a child with a higher rating than their left neighbor
        // receives more candies than that neighbor.
        for (int idx = 1; idx < plen; ++idx) {
            // If the current child's rating is greater than their left neighbor's rating,
            // they should receive one more candy than their left neighbor.
            if (ratings[idx] > ratings[idx-1]) {
                treats[idx] = treats[idx-1] + 1;
            }
        }
        
        // --- Right to left pass and sum ---
        // This pass ensures that a child with a higher rating than their right neighbor
        // receives more candies than that neighbor. It also calculates the total candies.
        
        // Initialize 'finalTreats' with the candies for the last child.
        // This is the starting point for summing up candies from right to left.
        int finalTreats = treats[plen-1];
        
        // Iterate from the second-to-last child back to the first child.
        for (int idx = plen - 2; idx >= 0; --idx) {
            // If the current child's rating is greater than their right neighbor's rating,
            // they *might* need more candies.
            if (ratings[idx] > ratings[idx+1]) {
                // We take the maximum of the candies already assigned (from the left-to-right pass)
                // and the candies required based on the right neighbor's rating plus one.
                // This handles cases where a child might have a higher rating than *both* neighbors.
                treats[idx] = std::max(treats[idx], treats[idx+1] + 1);
            }
            // Add the current child's determined candy count to the total.
            finalTreats += treats[idx];
        }
        
        // Return the total minimum number of candies required.
        return finalTreats;
    }
};
